### ISSUE TITLE
Shortcut for filtering results on a n:m relations

### ISSUE THREAD

**Issue by glamorous** [jonathangeiger/kohana-jelly#79]
I need a query to filter my results with their relationship.

Let use us the following example:

Model_Post
Model_Category

If I want all posts with category "PHP", then I could write a query or build the query with Jelly (and Kohana's query builder)

But isn't there a shortcut for this using "where" or something like that?

**Comment by ccazette** [comment]
That is an interesting question as I am going through something similar. Not sure it is exactly the same, but I think all in all it relates to the Field_Relationship consider the "name" property instead of an ID, so that if you query 
Jelly::select('post')->where('category', '=', 'php')->execute();
that would under the hood do :
Jelly::select('post')->where('category.category_name', '=', 'php')->execute();
if category_name is set as the "name" property in Model_Category.
But then that would enable to sort your query using that column, to filter using that column, or even to select only this column, instead of using the "column" property defined in the field, which is often an ID, not very relevant to sort with or filter out.

Hope what I am saying makes sense, but it's funny to see that this issue pops out at the very same time I am thinking about it...

**Comment by glamorous** [comment]
If it's not intended to have this as a feature request. Can someone help me then out to create such a thing with the custom query builder?

```
Jelly::select('post')->filterbycategory('php')->execute();
```

And then in the cusom builder I should be joining the trough table and the Category-table?

**Comment by jonathangeiger** [comment]
I think what you're asking could be accomplished with smart joins. Something like:

```
Jelly::select('post')->join('category');
```

This would automatically make the right JOIN based on the 'category' field's relationship type. After that, you can just do:

```
Jelly::select('post')->join('category')->where('category.name', '=', 'PHP')->execute();
```

This has the potential to be slightly confusing, however. In the join() method, you'd have to specify a field name for the post model, whereas in the where() clause, the 'category' part of 'category.name' is a model. So, for example, if your category field was named 'foo', it would be this:

```
Jelly::select('post')->join('foo')->where('category.name', '=', 'PHP')->execute();
```

That said, I think what ccazette proposed:

```
Jelly::select('post')->where('category', '=', 'php')->execute();
```

Is too much magic. I'd rather have a join() method like above. It's much clearer, more flexible, and more SQL-like without too much added verbosity.

It may be necessary to name the method auto_join() or something else, because I can see the potential for conflicts. For example, how is join() supposed to know whether you want to join a field or a model, if you have a field and model named the same thing?

There's some overlap with what's been discussed [here](http://github.com/jonathangeiger/kohana-jelly/issues#issue/27). I'm going to talk to banks about it and see if we might need to rethink that issue.

**Comment by jonathangeiger** [comment]
> Can someone help me then out to create such a thing with the custom query builder?

Sure thing. What kind of relationship is joining these two models?

**Comment by glamorous** [comment]
Model_Category

```
class Model_Category extends Jelly_Model
{
    public static function initialize(Jelly_Meta $meta)
    {
    $meta->fields(array(
        'id' => new Field_Primary,
        'uri' => new Field_Slug,
        'name' => new Field_String,
        'newsitems' => new Field_ManyToMany,
        'events' => new Field_ManyToMany,
    ));
    }
}
```

Model_Newsitem

```
class Model_Newsitem extends Jelly_Model
{
public static function initialize(Jelly_Meta $meta)
    {
    $meta->fields(array(
        'id' => new Field_Primary,
        'author' => new Field_BelongsTo(array(
            'foreign' => 'user',
        )),
        'event' => new Field_HasOne,
        'date' => new Field_Timestamp(array(
            'auto_now_create' => TRUE,
        )),
        'name' => new Field_String,
        'uri' => new Field_Slug,
        'body' => new Field_Text,
        'status' => new Field_Enum(array(
            'choices' => array('open','closed','pending'),
        )),
        'categories' => new Field_ManyToMany,
    ))
        ->load_with(array('author'))
        ->sorting(array('date' => 'DESC'));
    }
}
```

Enough information? Is this correct for the relationship?

**Comment by jonathangeiger** [comment]
> Enough information? Is this correct for the relationship?

Yep. The relationship seems fine to me. I guess you need a news item to be able to belong to many categories, right?

I'll get back to you in a bit about the proper JOIN.

**Comment by glamorous** [comment]
Jup, that's correct. If I have the correct "solution", I can port it to my other models too... Thanks in advance!

**Comment by jonathangeiger** [comment]
Here's your solution glamorous. This is also a perfect example for why auto_join() would be nice:

```
class Model_Builder_Newsitem extends Jelly_Builder
{
    public function find_by_category($category)
    {
        return $this->join('categories_newsitems', 'INNER')
                    ->on('newsitem.id', '=', 'categories_newsitems.newsitem_id')
                    ->join('category', 'INNER')
                    ->on('category.id', '=', 'categories_newsitems.category_id')
                    ->where('category.name', '=', $category);
    }
}
```

That should get you there. Let me know that it works. There might be a few typos with the field/model names.

**Comment by glamorous** [comment]
Thanks jonathan, that function works very fine. The only change I did was 'category.uri' instead of 'category.name' because I wanted to use the slug-version of the category from in the url...

**Comment by ccazette** [comment]
@Jon: Yes, I was clearly out of my mind, forgot you have to join with categories first ! Too much magic obviously...

So yes, an auto_join() method would make a lot of sense IMO, more than for_key() because way more flexible. I already suggested this last month and discussed it with banks, perhaps some toughts could still be useful :
http://forum.kohanaphp.com/comments.php?DiscussionID=4713&page=3#Item_17
